File: ios/barikoi_trace_flutter/Sources/barikoi_trace_flutter/LogStreamHandler.swift
import Flutter
import Foundation

import BarikoiTrace

/// Backs `barikoi_trace_flutter/logs`.
///
/// Three things make this more than a straight adapter:
///
///  1. **`TraceManager.logListener` is `weak`.** `BarikoiTrace.setLogListener`
///     assigns straight into it, so nothing in the SDK keeps this object
///     alive. The plugin instance holds it strongly (and the plugin instance
///     is itself retained by `addMethodCallDelegate`), which is the only
///     reason logs keep arriving after `register(with:)` returns.
///  2. **A replay buffer.** The launch-time and `initialize`-time
///     configuration warnings are produced before Dart has had any chance to
///     subscribe — they are, in fact, exactly the diagnostics an integrator
///     needs most. They are buffered and flushed to the first subscriber.
///  3. **Everything happens on the main thread.** The SDK's log callback fires
///     on whatever thread produced the line, including from inside SDK code
///     the plugin itself called. Hopping to main before touching the sink
///     satisfies the channel's threading rule and, as a side effect, makes the
///     buffer/sink pair single-threaded so it needs no lock.
final class LogStreamHandler: NSObject, FlutterStreamHandler, TraceLogListener {

    /// Roughly a screenful of scrollback. Old entries are dropped, not the new
    /// ones — a full buffer means the interesting lines are the recent ones.
    private static let bufferCapacity = 200

    /// Main-thread only. See the class comment.
    private var buffer: [[String: Any]] = []
    private var sink: FlutterEventSink?

    // MARK: - TraceLogListener

    func onLog(level: String, tag: String, message: String) {
        emit(level: level, tag: tag, message: message)
    }

    // MARK: - Plugin-authored entries

    /// Emits a line the plugin itself produced (config warnings, AppDelegate
    /// ordering diagnostics).
    ///
    /// These deliberately do **not** go through `TraceManager.log`, for two
    /// reasons: that method drops everything while `setLoggingEnabled(false)`
    /// — which is the state on a fresh install, before the SDK's own
    /// `initialize` turns logging on — and it would then call back into this
    /// object as the registered listener, duplicating every entry.
    func emit(level: String, tag: String, message: String) {
        let entry = Codecs.encodeLogEntry(level: level, tag: tag, message: message)
        if Thread.isMainThread {
            deliver(entry)
        } else {
            DispatchQueue.main.async { [weak self] in self?.deliver(entry) }
        }
    }

    /// Main thread only.
    private func deliver(_ entry: [String: Any]) {
        if let sink = sink {
            sink(entry)
        } else {
            if buffer.count >= LogStreamHandler.bufferCapacity {
                buffer.removeFirst(buffer.count - LogStreamHandler.bufferCapacity + 1)
            }
            buffer.append(entry)
        }
    }

    // MARK: - FlutterStreamHandler

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        sink = events
        let replay = buffer
        buffer.removeAll()
        for entry in replay {
            events(entry)
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        // Back to buffering. A Dart-side hot restart cancels and re-listens,
        // and the lines produced in between are still worth showing.
        sink = nil
        return nil
    }
}

File: android/src/main/kotlin/com/barikoi/barikoitrace/flutter/LogStreamHandler.kt
package com.barikoi.barikoitrace.flutter

import android.os.Handler
import android.os.Looper
import com.barikoi.barikoitrace.BarikoiTrace
import io.flutter.plugin.common.EventChannel

/**
 * `barikoi_trace_flutter/logs` — one `TraceLogEntry` map per native log line.
 *
 * Installed as the single `BarikoiTrace.TraceLogListener` for the process and
 * fanned out to whatever is currently listening on the event channel.
 *
 * Why the ring buffer: the most useful log lines the SDK produces are the
 * `TraceConfig.warnings` it emits from inside `initialize()` — empty API key,
 * empty broker credentials, plaintext `mqttUrl`, non-HTTPS `baseUrl`. Those
 * fire before any realistic Dart subscriber exists (the host has to call
 * `initialize` to have a plugin worth listening to). Without a replay they are
 * simply lost, which is the opposite of what a diagnostics channel is for. So
 * entries accumulate while nobody is listening, capped at [MAX_BUFFERED], and
 * are handed to the next subscriber before live traffic.
 *
 * Threading, per WIRE_CONTRACT §7.2: `onLog` can fire from any thread — an
 * OkHttp callback, the foreground service, the MQTT client — and it can fire
 * re-entrantly from inside SDK code the plugin itself called. Everything
 * therefore hops to the main thread, and delivery runs as a drain loop guarded
 * by [draining], so a log line emitted from inside `EventSink.success` is
 * queued and flushed by the loop already running rather than recursing.
 */
internal class LogStreamHandler : EventChannel.StreamHandler, BarikoiTrace.TraceLogListener {

    private val mainHandler = Handler(Looper.getMainLooper())

    private val bufferLock = Any()
    private val buffered = ArrayDeque<Map<String, Any?>>()

    /** Touched on the main thread only. */
    private var sink: EventChannel.EventSink? = null
    private var draining = false

    // --- TraceLogListener ---

    override fun onLog(level: String, tag: String, message: String) {
        // Stamped here, at callback time: the native callback carries no
        // timestamp of its own (WIRE_CONTRACT §4.5).
        val entry = Codecs.encodeLogEntry(level, tag, message, System.currentTimeMillis())

        synchronized(bufferLock) {
            buffered.addLast(entry)
            while (buffered.size > MAX_BUFFERED) {
                buffered.removeFirst()
            }
        }

        if (Looper.myLooper() == Looper.getMainLooper()) {
            drain()
        } else {
            mainHandler.post { drain() }
        }
    }

    // --- EventChannel.StreamHandler ---

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        // Replays whatever accumulated before this subscriber arrived —
        // initialize-time warnings above all — then continues live.
        drain()
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }

    /**
     * Called when the engine detaches. Drops the sink and the backlog: the next
     * engine gets a fresh handler, and replaying a previous engine's log into
     * it would be misleading.
     */
    fun dispose() {
        sink = null
        synchronized(bufferLock) { buffered.clear() }
    }

    // --- Delivery ---

    /** Main thread only. */
    private fun drain() {
        if (draining) return
        val target = sink ?: return

        draining = true
        try {
            while (true) {
                val next = synchronized(bufferLock) { buffered.removeFirstOrNull() } ?: break
                try {
                    target.success(next)
                } catch (_: Throwable) {
                    // The sink is gone or the messenger is shutting down. The
                    // log channel must never fail the stream (§7.2), and it must
                    // never log from inside the sink, so this is swallowed on
                    // purpose.
                    sink = null
                    break
                }
            }
        } finally {
            draining = false
        }
    }

    private companion object {
        /**
         * Roughly one screen of scrollback. Large enough to hold every
         * initialize-time warning plus the first burst of session logs, small
         * enough that an app which never listens cannot grow it without bound.
         */
        const val MAX_BUFFERED = 200
    }
}
